### PR TITLE
Rootdevicehints added to BMH spec

### PR DESCRIPTION
**Pull request by mikkosest**
THIS PR WILL BE DEPRICATED!

### CODE REVIEW

**Comment by dhellmann** [comment]
This change exposes all of the root device hinting features of Ironic. I'm not entirely sure we need all of them, do we? Can we limit the set to just the "hints" that let the user specify an exact device, and leave out things like "rotational", "size", and "vendor"?

**Comment by dhellmann** [comment]
> This change exposes all of the root device hinting features of Ironic. I'm not entirely sure we need all of them, do we? Can we limit the set to just the "hints" that let the user specify an exact device, and leave out things like "rotational", "size", and "vendor"?

Normally we would have that sort of discussion on a design document in the metal3-docs repo. Perhaps you can write one of those for us to use to discuss the direction we want to take, and we can pause the implementation work until we agree there?

**Comment by mikkosest** [comment]
> > This change exposes all of the root device hinting features of Ironic. I'm not entirely sure we need all of them, do we? Can we limit the set to just the "hints" that let the user specify an exact device, and leave out things like "rotational", "size", and "vendor"?
> 
> Normally we would have that sort of discussion on a design document in the metal3-docs repo. Perhaps you can write one of those for us to use to discuss the direction we want to take, and we can pause the implementation work until we agree there?


OK, I'will make a design doc. We would need at least WWN parts.

**Comment by mikkosest** [comment]
> > This change exposes all of the root device hinting features of Ironic. I'm not entirely sure we need all of them, do we? Can we limit the set to just the "hints" that let the user specify an exact device, and leave out things like "rotational", "size", and "vendor"?
> 
> Normally we would have that sort of discussion on a design document in the metal3-docs repo. Perhaps you can write one of those for us to use to discuss the direction we want to take, and we can pause the implementation work until we agree there?

I made a drafty proposal on https://github.com/metal3-io/metal3-docs/pull/76/files.
 Could you please check? @dhellmann @zaneb @yrobla @stbenjam @maelk

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go [hunk 1]
```diff
@@ -24,6 +24,40 @@ const (
...
+	Serial string `json:"serial,omitempty"`
+
+	// Size of the device in GiB
+	Size int `json:"size,omitempty"`
+
+	// Unique storage identifier
+	WWN string `json:"wwn,omitempty"`
+
+	// Unique storage identifier with the vendor extension appended
+	WWNwithExtention string `json:"wwnwithextention,omitempty"`
```
Comment (zaneb) [inline pkg/apis/metal3/v1alpha1/baremetalhost_types.go]: Should be `WWNWithExtension` and json `wwnWithExtension` for consistency with HardwareDetails

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go [hunk 2]
```diff
@@ -24,6 +24,40 @@ const (
...
+
+	// Uunique vendor storage identifier
+	WWNvendorExtention string `json:"wwnvendorextention,omitempty"`
```
Comment (zaneb) [inline pkg/apis/metal3/v1alpha1/baremetalhost_types.go]: Should be `WWNVendorExtension` and json `wwnVendorExtension`

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go [hunk 3]
```diff
@@ -24,6 +24,40 @@ const (
...
+	// Size of the device in GiB
+	Size int `json:"size,omitempty"`
+
+	// Unique storage identifier
+	WWN string `json:"wwn,omitempty"`
+
+	// Unique storage identifier with the vendor extension appended
+	WWNwithExtention string `json:"wwnwithextention,omitempty"`
+
+	// Uunique vendor storage identifier
```
Comment (zaneb) [inline pkg/apis/metal3/v1alpha1/baremetalhost_types.go]: Uunique spelling ;)

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go [hunk 4]
```diff
@@ -24,6 +24,40 @@ const (
...
+	WWNvendorExtention string `json:"wwnvendorextention,omitempty"`
+
+	// Device rotational type
+	Rotational string `json:"rotational,omitempty"`
```
Comment (zaneb) [inline pkg/apis/metal3/v1alpha1/baremetalhost_types.go]: Should be a bool

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go [hunk 5]
```diff
@@ -24,6 +24,40 @@ const (
...
+	Model string `json:"model,omitempty"`
+
+	// Device vendor
+	Vendor string `json:"vendor,omitempty"`
+
+	// Disk serial number
+	Serial string `json:"serial,omitempty"`
+
+	// Size of the device in GiB
+	Size int `json:"size,omitempty"`
```
Comment (zaneb) [inline pkg/apis/metal3/v1alpha1/baremetalhost_types.go]: Why GiB? Elsewhere (e.g. HardwareDetails) we always report storage sizes in bytes.
Also, we should include the units in the name.

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go [hunk 6]
```diff
@@ -24,6 +24,40 @@ const (
 	PausedAnnotation = "baremetalhost.metal3.io/paused"
 )
 
+// RootDeviceHints holds the hints for specifying the storage location
+// for the root filesystem for the image.
+type RootDeviceHints struct {
```
Comment (zaneb) [inline pkg/apis/metal3/v1alpha1/baremetalhost_types.go]: This is mostly identical to the existing `Storage` struct, and many of the differences are not improvements. It would be worth investigating the feasibility of just reusing that struct.

File: deploy/crds/metal3.io_baremetalhosts_crd.yaml [hunk 7]
```diff
@@ -51,12 +51,12 @@ spec:
         apiVersion:
           description: 'APIVersion defines the versioned schema of this representation
             of an object. Servers should convert recognized schemas to the latest
-            internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources'
+            internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#resources'
```
Comment (yrobla) [inline deploy/crds/metal3.io_baremetalhosts_crd.yaml]: this looks unrelated

File: deploy/crds/metal3.io_baremetalhosts_crd.yaml [hunk 8]
```diff
@@ -159,6 +159,42 @@ spec:
...
+                vendor:
+                  description: Device vendor
+                  type: string
+                wwn:
+                  description: Unique storage identifier
+                  type: string
+                wwnvendorextention:
+                  description: Uunique vendor storage identifier
+                  type: string
+                wwnwithextention:
```
Comment (yrobla) [inline deploy/crds/metal3.io_baremetalhosts_crd.yaml]: wwn_with_extension

File: deploy/crds/metal3.io_baremetalhosts_crd.yaml [hunk 9]
```diff
@@ -159,6 +159,42 @@ spec:
...
+                size:
+                  description: Size of the device in GiB
+                  type: integer
+                vendor:
+                  description: Device vendor
+                  type: string
+                wwn:
+                  description: Unique storage identifier
+                  type: string
+                wwnvendorextention:
```
Comment (yrobla) [inline deploy/crds/metal3.io_baremetalhosts_crd.yaml]: wwn_vendor_extension

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go (hunk 4 above)
Comment (yrobla) [reply pkg/apis/metal3/v1alpha1/baremetalhost_types.go]: I guess it depends on the content that the field will hold. Do we want to add just values here, like "true" ? Or do we want to support the full operator of root device hints? Like =true, !=true , etc... When looking at https://docs.openstack.org/ironic/rocky/install/advanced.html , you can see that there is a full list of logical operators that can be supported

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go [hunk 10]
```diff
@@ -559,6 +597,10 @@ func (host *BareMetalHost) HardwareProfile() string {
 // SetHardwareProfile updates the hardware profile name and returns
 // true when a change is made or false when no change is made.
 func (host *BareMetalHost) SetHardwareProfile(name string) (dirty bool) {
+	if host.Status.HardwareProfile != name {
```
Comment (yrobla) [inline pkg/apis/metal3/v1alpha1/baremetalhost_types.go]: why is that for? Isn't the same as lower code?

File: pkg/apis/metal3/v1alpha1/zz_generated.openapi.go [hunk 11]
```diff
@@ -24,14 +24,14 @@ func schema_pkg_apis_metal3_v1alpha1_BareMetalHost(ref common.ReferenceCallback)
 				Properties: map[string]spec.Schema{
 					"kind": {
 						SchemaProps: spec.SchemaProps{
-							Description: "Kind is a string value representing the REST resource this object represents. Servers may infer this from the endpoint the client submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds",
+							Description: "Kind is a string value representing the REST resource this object represents. Servers may infer this from the endpoint the client submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#types-kinds",
```
Comment (yrobla) [inline pkg/apis/metal3/v1alpha1/zz_generated.openapi.go]: looks unrelated

File: deploy/crds/metal3.io_baremetalhosts_crd.yaml (hunk 8 above)
Comment (stbenjam) [reply deploy/crds/metal3.io_baremetalhosts_crd.yaml]: It should always be camel case in k8s, wwwWithExtension as @zaneb mentioned above

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go (hunk 1 above)
Comment (mikkosest) [reply pkg/apis/metal3/v1alpha1/baremetalhost_types.go]: Done

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go (hunk 2 above)
Comment (mikkosest) [reply pkg/apis/metal3/v1alpha1/baremetalhost_types.go]: Done

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go (hunk 3 above)
Comment (mikkosest) [reply pkg/apis/metal3/v1alpha1/baremetalhost_types.go]: Fixed

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go (hunk 4 above)
Comment (mikkosest) [reply pkg/apis/metal3/v1alpha1/baremetalhost_types.go]: Done, it seems that gopher module still takes string map as input so I need to do conversion to string in code, right?

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go (hunk 5 above)
Comment (mikkosest) [reply pkg/apis/metal3/v1alpha1/baremetalhost_types.go]: Changes to SizeGigaBytes. I would like to keep GS, as ironic takes GB as input, so I would avoid unit conversion. OK?

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go (hunk 6 above)
Comment (mikkosest) [reply pkg/apis/metal3/v1alpha1/baremetalhost_types.go]: Related to above size issue and proposal to use subset in metal3-doc discussion I would like to keep separate structs.  OK?

File: deploy/crds/metal3.io_baremetalhosts_crd.yaml (hunk 7 above)
Comment (mikkosest) [reply deploy/crds/metal3.io_baremetalhosts_crd.yaml]: Fixed

File: deploy/crds/metal3.io_baremetalhosts_crd.yaml (hunk 8 above)
Comment (mikkosest) [reply deploy/crds/metal3.io_baremetalhosts_crd.yaml]: Fixed

File: deploy/crds/metal3.io_baremetalhosts_crd.yaml (hunk 9 above)
Comment (mikkosest) [reply deploy/crds/metal3.io_baremetalhosts_crd.yaml]: Changed to camel

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go (hunk 10 above)
Comment (mikkosest) [reply pkg/apis/metal3/v1alpha1/baremetalhost_types.go]: Removed

File: pkg/apis/metal3/v1alpha1/zz_generated.openapi.go (hunk 11 above)
Comment (mikkosest) [reply pkg/apis/metal3/v1alpha1/zz_generated.openapi.go]: Fiexd

File: deploy/crds/metal3.io_baremetalhosts_crd.yaml [hunk 12]
```diff
@@ -159,6 +159,42 @@ spec:
             online:
               description: Should the server be online?
               type: boolean
+            rootDeviceHints:
+              description: What user defined root device hints is used for this host?
```
Comment (yrobla) [inline deploy/crds/metal3.io_baremetalhosts_crd.yaml]: is/are

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go [hunk 13]
```diff
@@ -24,6 +24,40 @@ const (
...
+	Model string `json:"model,omitempty"`
+
+	// Device vendor
+	Vendor string `json:"vendor,omitempty"`
+
+	// Disk serial number
+	Serial string `json:"serial,omitempty"`
+
+	// Size of the device in Gigabytes
+	SizeGigaBytes int `json:"sizeBytes,omitempty"`
```
Comment (yrobla) [inline pkg/apis/metal3/v1alpha1/baremetalhost_types.go]: why not just Size, and document that needs to be gigabyte?

File: deploy/crds/metal3.io_baremetalhosts_crd.yaml [hunk 14]
```diff
@@ -159,6 +159,42 @@ spec:
...
+                model:
+                  description: Device identifier
+                  type: string
+                rotational:
+                  description: Device rotational type
+                  type: boolean
+                serial:
+                  description: Disk serial number
+                  type: string
+                sizeBytes:
```
Comment (yrobla) [inline deploy/crds/metal3.io_baremetalhosts_crd.yaml]: why not just size, and document that you need to specify on gigabytes? Also, it's confusing as you say "bytes"

**Comment by derekhiggins** [comment]
I've pushed up a related PR for masters to the installer repository
https://github.com/openshift/installer/pull/3348

It suggests a simpler roothint pass through approach, would it be worth considering and allowing ironic to do the validation?

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go [hunk 15]
```diff
@@ -24,6 +24,40 @@ const (
...
```
Comment (dhellmann) [inline pkg/apis/metal3/v1alpha1/baremetalhost_types.go:46]:
```suggestion
	SizeGigabytes int `json:"sizeGigabytes,omitempty"`
```

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go (hunk 13 above)
Comment (dhellmann) [reply pkg/apis/metal3/v1alpha1/baremetalhost_types.go:46]: These fields end up in the API, and we want users of the API to easily know what sorts of values to fill in. The size for storage is in bytes, but this field uses GB, so being explicit is better.

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go [hunk 16]
```diff
@@ -159,6 +193,10 @@ type BareMetalHostSpec struct {
 	// automatically determine the profile.
 	HardwareProfile string `json:"hardwareProfile,omitempty"`
 
+	// What user defined root device hints is used for this host? User can optionally
+	// define the content of the root device hints
+	RootDeviceHints RootDeviceHints `json:"rootDeviceHints,omitempty"`
```
Comment (dhellmann) [inline pkg/apis/metal3/v1alpha1/baremetalhost_types.go:198]: I think in order to be optional this needs to be a pointer field.

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go [hunk 17]
```diff
@@ -426,6 +464,9 @@ type BareMetalHostStatus struct {
 	// The hardware discovered to exist on the host.
 	HardwareDetails *HardwareDetails `json:"hardware,omitempty"`
 
+	// The RootDevicehints set by the user
+	RootDeviceHints RootDeviceHints `json: "rootDeviceHints,omitempty"`
```
Comment (dhellmann) [inline pkg/apis/metal3/v1alpha1/baremetalhost_types.go:468]: I think the point of this was to record what was used on the last provisioning operation. If so, maybe it should go into the Provisioning struct?

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go (hunk 15 above)
Comment (zaneb) [reply pkg/apis/metal3/v1alpha1/baremetalhost_types.go]: This would be the only place we use Gigabytes instead of Gibibytes.

**Comment by maelk** [comment]
I'll close this pull request as @dhellman 's pull request was merged. Please re-open if needed.
/close
